src/orange/universes/SimpleUnitTracker.hh

```diff
@@ -90,47 +99,84 @@ CELER_FUNCTION SimpleUnitTracker::SimpleUnitTracker(const ParamsRef& params)
 
 //---------------------------------------------------------------------------//
 /*!
- * Find the local cell and possibly surface ID.
+ * Find the local volume from a position.
  *
- * This function is valid for initialization from a point, *and* for
- * initialization across a boundary.
+ * To avoid edge cases and inconsistent logical/physical states, it is
+ * prohibited to initialize from an arbitrary point directly onto a surface.
  */
 CELER_FUNCTION auto
 SimpleUnitTracker::initialize(const LocalState& state) const -> Initialization
 {
     CELER_EXPECT(params_);
+    CELER_EXPECT(!state.surface && !state.volume);
 
     detail::SenseCalculator calc_senses(
         Surfaces{params_.surfaces}, state.pos, state.temp_sense);
 
+    // Loop over all volumes (TODO: use BVH)
     for (VolumeId volid : range(VolumeId{params_.volumes.size()}))
     {
-        if (state.surface && volid == state.volume)
+        VolumeView vol{params_.volumes, volid};
+
+        // Calculate the local senses, and see if we're inside.
+        auto logic_state = calc_senses(vol);
+
+        // Evalulate whether the senses are "inside" the volume
+        if (!detail::LogicEvaluator(vol.logic())(logic_state.senses))
         {
-            // Cannot cross surface into the same cell
+            // State is *not* inside this volume: try the next one
             continue;
         }
+        if (logic_state.face)
+        {
+            // Initialized on a boundary in this volume but wasn't known
+            // to be crossing a surface. Fail safe by letting the multi-level
+            // tracking geometry bump and try again.
```

Comment:
Can you elaborate on this? What happens if the position is right on a boundary?

Reply:
This is a feature of the SCALE CPU version that I'll be adding to this version eventually. I came to the conclusion that the safest way to avoid logical errors and to reason easily about the program, we need to bump particles that are initialized on a boundary:
- Coincident surface between daughter and parent (these are eliminated in most cases by the daughter universe construction so not something that generally happens)
- Crossing a boundary onto an edge (usually only happens when ray tracing)
- Initializing from a starting source onto a boundary (usually only ray tracing or poor user choices)

For now the transport will just fail if we hit an edge wrong, and we just have to hope that there's no material boundary where primaries are created.

Reply:
Thanks for the explanation!